packages/messaging/index.ts

```diff
@@ -19,20 +19,29 @@ import {
   _FirebaseNamespace,
   FirebaseServiceFactory
 } from '@firebase/app-types/private';
+
 import { SWController } from './src/controllers/sw-controller';
 import { WindowController } from './src/controllers/window-controller';
+import { ERROR_CODES, errorFactory } from './src/models/errors';
 
 import * as types from '@firebase/messaging-types';
 
 export function registerMessaging(instance: _FirebaseNamespace): void {
   const messagingName = 'messaging';
   const factoryMethod: FirebaseServiceFactory = app => {
     if (self && 'ServiceWorkerGlobalScope' in self) {
+      // Running in ServiceWorker context
       return new SWController(app);
-    }
+    } else {
+      // Assume we are in the window context.
+
+      // Check that the required API's are available
+      if (!isWindowControllerSupported()) {
+        throw errorFactory.create(ERROR_CODES.UNSUPPORTED_BROWSER);
```

Comment:
Won't this cause an app to completely fail just by loading the library?

Reply:
No. It only throws when `firebase.messaging()` is called. You can still import '@firebase/messaging' or the complete firebase bundle without any errors.

I think that if a user calls `firebase.messaging()` to get the messaging service, they should receive an error instead of getting a service controller that will throw as soon as getToken is called. WDYT?

Reply:
It's going to break current developers and should be done with a major version bump.

Reply:
So you're okay with it as long as we bump the version? We can release this with v5.